File: 098_ValidateBinarySearchTree.cpp

//59.87%
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if(!root)
			return true;
		bool flagLeft(false);
		bool flagRight(false);
		if(!root -> left)
			flagLeft = true;
		else{
			TreeNode* temp(root -> left);
			while(temp -> right) temp = temp -> right;
			if(temp -> val < root -> val && isValidBST(root -> left))
				flagLeft = true;
		}
		
		if(!root -> right)
			flagRight = true;
		else{
			TreeNode* temp(root -> right);
			while(temp -> left) temp = temp -> left;
			if(temp -> val > root -> val && isValidBST(root -> right))
				flagRight = true;
		}
		return flagLeft && flagRight;
    }
};



//59.87% -> 9.28%%
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if(!root)
			return true;
		bool flag(false);
		if(!root -> left)
			flag = true;
		else{
			TreeNode* temp(root -> left);
			while(temp -> right) temp = temp -> right;
			if(temp -> val < root -> val && isValidBST(root -> left))
				flag = true;
		}
		if(!flag)
			return false;
		flag = false;
		if(!root -> right)
			flag = true;
		else{
			TreeNode* temp(root -> right);
			while(temp -> left) temp = temp -> left;
			if(temp -> val > root -> val && isValidBST(root -> right))
				flag = true;
		}
		return flag;
    }
};

//BST inorder traversal
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        TreeNode* prev = NULL;
        return validate(root, prev);
    }
    bool validate(TreeNode* node, TreeNode* &prev) {
        if (node == NULL) return true;
        if (!validate(node->left, prev)) return false;
        if (prev != NULL && prev->val >= node->val) return false;
        prev = node;
        return validate(node->right, prev);
    }
};

//A java iterative way
public boolean isValidBST(TreeNode root) {
   if (root == null) return true;
   Stack<TreeNode> stack = new Stack<>();
   TreeNode pre = null;
   while (root != null || !stack.isEmpty()) {
      while (root != null) {
         stack.push(root);
         root = root.left;
      }
      root = stack.pop();
      if(pre != null && root.val <= pre.val) return false;
      pre = root;
      root = root.right;
   }
   return true;
}

//a clever recursive function and it's thinking looks like mine
bool isValidBST(TreeNode* root) {
    return isValidBST(root, NULL, NULL);
}

bool isValidBST(TreeNode* root, TreeNode* minNode, TreeNode* maxNode) {
    if(!root) return true;
    if(minNode && root->val <= minNode->val || maxNode && root->val >= maxNode->val)
        return false;
    return isValidBST(root->left, minNode, root) && isValidBST(root->right, root, maxNode);
}

File: 098_ValidateBinarySearchTree.py
# 40.96%%
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        
        flagLeft, flagRight = False, False
        if not root.left:
            flagLeft = True
        else:
            temp = root.left
            while temp.right: temp = temp.right
            if temp.val < root.val and self.isValidBST(root.left):
                flagLeft = True
        if not flagLeft:
            return False
        if not root.right:
            flagRight = True
        else:
            temp = root.right
            while temp.left: temp = temp.left
            if temp.val > root.val and self.isValidBST(root.right):
                flagRight = True
        return flagRight
        
###Inorder traversal
class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return self.isBSTHelper(root, [])
        
    def isBSTHelper(self, root, prev=[]):
        if not root:
            return True
        if not self.isBSTHelper(root.left, prev):
            return False
        if not prev:
            prev.append(root.val)
        elif root.val <= prev[0]:
            return False
        prev[0] = root.val
        if not self.isBSTHelper(root.right, prev):
            return False
        return True
		
		
class Solution(object):
    def isValidBST(self, root, lessThan = float('inf'), largerThan = float('-inf')):
        if not root:
            return True
        if root.val <= largerThan or root.val >= lessThan:
            return False
        return self.isValidBST(root.left, min(lessThan, root.val), largerThan) and \
               self.isValidBST(root.right, lessThan, max(root.val, largerThan))
      
        

                              